clients/src/test/java/org/apache/kafka/common/metrics/stats/HistogramTest.java

```diff
@@ -33,29 +33,103 @@
 
     @Test
     public void testHistogram() {
-        BinScheme scheme = new ConstantBinScheme(12, -5, 5);
+        BinScheme scheme = new ConstantBinScheme(10, -5, 5);
         Histogram hist = new Histogram(scheme);
         for (int i = -5; i < 5; i++)
             hist.record(i);
         for (int i = 0; i < 10; i++)
-            assertEquals(scheme.fromBin(i + 1), hist.value(i / 10.0 + EPS), EPS);
+            assertEquals(scheme.fromBin(i), hist.value(i / 10.0 + EPS), EPS);
     }
 
     @Test
     public void testConstantBinScheme() {
         ConstantBinScheme scheme = new ConstantBinScheme(5, -5, 5);
         assertEquals("A value below the lower bound should map to the first bin", 0, scheme.toBin(-5.01));
         assertEquals("A value above the upper bound should map to the last bin", 4, scheme.toBin(5.01));
-        assertEquals("Check boundary of bucket 1", 1, scheme.toBin(-5));
-        assertEquals("Check boundary of bucket 4", 4, scheme.toBin(5));
-        assertEquals("Check boundary of bucket 3", 3, scheme.toBin(4.9999));
-        checkBinningConsistency(new ConstantBinScheme(4, 0, 5));
+        assertEquals("Check boundary of bucket 0", 0, scheme.toBin(-5.0001));
+        assertEquals("Check boundary of bucket 0", 0, scheme.toBin(-5.0000));
+        assertEquals("Check boundary of bucket 0", 0, scheme.toBin(-4.99999));
+        assertEquals("Check boundary of bucket 0", 0, scheme.toBin(-3.00001));
+        assertEquals("Check boundary of bucket 1", 1, scheme.toBin(-3));
+        assertEquals("Check boundary of bucket 1", 1, scheme.toBin(-1.00001));
+        assertEquals("Check boundary of bucket 2", 2, scheme.toBin(-1));
+        assertEquals("Check boundary of bucket 2", 2, scheme.toBin(0.99999));
+        assertEquals("Check boundary of bucket 3", 3, scheme.toBin(1));
+        assertEquals("Check boundary of bucket 3", 3, scheme.toBin(2.99999));
+        assertEquals("Check boundary of bucket 4", 4, scheme.toBin(3));
+        assertEquals("Check boundary of bucket 4", 4, scheme.toBin(4.9999));
+        assertEquals("Check boundary of bucket 4", 4, scheme.toBin(5.000));
+        assertEquals("Check boundary of bucket 4", 4, scheme.toBin(5.001));
+        assertEquals(Float.NEGATIVE_INFINITY, scheme.fromBin(-1), 0.001d);
+        assertEquals(Float.POSITIVE_INFINITY, scheme.fromBin(5), 0.001d);
+        assertEquals(-5.0, scheme.fromBin(0), 0.001d);
+        assertEquals(-3.0, scheme.fromBin(1), 0.001d);
+        assertEquals(-1.0, scheme.fromBin(2), 0.001d);
+        assertEquals(1.0, scheme.fromBin(3), 0.001d);
+        assertEquals(3.0, scheme.fromBin(4), 0.001d);
+        checkBinningConsistency(scheme);
```

Comment:
Of course, it is possible to simplify these tests to use a loop, passing the output of `fromBin()` into `toBin()` and just checking they are consistent. That is what `checkBinningConsistency` does, and it like the name suggests it only checks that the two methods are consistent, not that the bins are computed correctly. IMO it is far clearer and more complete to use literal values and to check the range of the bins. I am intentionally using literals with different precision.